src/python/pants/cache/local_artifact_cache.py

```diff
@@ -65,34 +66,45 @@ def store_and_use_artifact(self, cache_key, src, results_dir=None):
         tmp.write(chunk)
       tmp.close()
       tarball = self._store_tarball(cache_key, tmp.name)
-      artifact = self._artifact(tarball)
+      return self.extract_artifact_from_tarball(tarball, results_dir)
 
-      # NOTE(mateo): The two clean=True args passed in this method are likely safe, since the cache will by
-      # definition be dealing with unique results_dir, as opposed to the stable vt.results_dir (aka 'current').
-      # But if by chance it's passed the stable results_dir, safe_makedir(clean=True) will silently convert it
-      # from a symlink to a real dir and cause mysterious 'Operation not permitted' errors until the workdir is cleaned.
-      if results_dir is not None:
-        safe_mkdir(results_dir, clean=True)
+  def extract_artifact_from_tarball(self, tarball, results_dir=None):
+    """Extract an artifact from a given tarball.
 
+    :param tarball: Path to the tarball to extract.
+    :param str results_dir: The path to the expected destination of the artifact extraction,
+      which will be cleared before extraction. This path must be under artifact_root, and all
+      paths in the artifact must be under this results_dir.
+    :return: True iff the extraction succeeded.
+    :rtype: bool
+    """
+    # Note: If results_dir is a symlink we operate on the underlying dir it references.
+    # Otherwise, safe_makedir(clean=True) will silently convert it from a symlink to a real dir.
```

Comment:
We no longer use `safe_makedir`, so the comment needs updating.

But also: we might want to clear up the pre/post conditions for the results_dir: in particular, if an artifact tarfile is empty, this method should probably still guarantee that an (empty) results directory is created (which was the effect of `safe_mkdir` before).

Finally, it looks like this `results_dir = os.path.realpath(results_dir)` conversion would need to happen inside of a check for `None`, since it may not exist.